src/util/misc.c

```diff
@@ -111,14 +111,24 @@ int misc_memfd(const char *name, unsigned int uflags, unsigned int useals) {
         }
 
         /*
-         * If we ended up passing `MFG_NOEXEC_SEAL` to the kernel, the kernel
+         * If we ended up passing `MFG_NOEXEC_SEAL` to the kernel, older kernels
          * will implicitly enable sealing. This is very unfortunate, so we
          * revert this if the caller did not explicitly allow it. To disable
          * sealing, simply set `F_SEAL_SEAL`, which is also what the kernel
          * does.
          */
-        if ((flags & MISC_MFD_NOEXEC_SEAL) && !(flags & MISC_MFD_ALLOW_SEALING))
-                seals |= MISC_F_SEAL_SEAL;
+        if ((flags & MISC_MFD_NOEXEC_SEAL) && !(flags & MISC_MFD_ALLOW_SEALING)) {
+                r = fcntl(fd, MISC_F_GET_SEALS);
+                if (r < 0)
+                        return error_origin(-errno);
+
+                /*
+                 * Newer kernels do not implicitly allow sealing on `MFG_NOEXEC_SEAL`,
+                 * so only try to set `F_SEAL_SEAL` if not already set by the kernel.
+                 */
+                if (!(r & MISC_F_SEAL_SEAL))
+                        seals |= MISC_F_SEAL_SEAL;
+        }
```

Comment:
How about you leave this part as it was and instead change the code below to this:

```
if (seals) {
        r = misc_memfd_get_seals(fd);
        if (r < 0)
                return error_fold(r);

        if (seals & ~r) {
                r = misc_memfd_add_seals(fd, seals);
                if (r)
                        return error_fold(r);
        }
}
```

This moves the `MISC_F_GET_SEALS` call down a bit, but safely works around the problem of the kernel not allowing F_SEAL_SEAL if F_SEAL_SEAL is already set.

Reply:
Yeah, I know, but I think that is fine. It is now 3 instead of 2 syscalls in that particular path (i.e., creating a memfd with initial seals, which is nothing we do, anyway). If we decide to optimize this path, we can always create fast-paths later. I prefer if the default is the "works for all" solution.